semgrep-core/src/fixing/Autofix.ml

```diff
@@ -0,0 +1,70 @@
+(* Nat Mote
+ *
+ * Copyright (C) 2019-2022 r2c
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+open Common
+
+(******************************************************************************)
+(* Main module for AST-based autofix. This module will attempt to synthesize a
+ * fix based a rule's fix pattern and the match's metavariable bindings. *)
+(******************************************************************************)
+
+let parse_pattern_opt lang pattern =
+  try Some (Parse_pattern.parse_pattern lang pattern) with
+  | Timeout _ as e -> Exception.catch_and_reraise e
+  | _ -> None
+
+(* Check whether the proposed fix results in syntactically valid code *)
+let validate_fix lang text =
+  (* TODO Apply fix to original text and try parsing the whole program *)
+  (* Since the fix is likely a program fragment, and not a valid top level
+   * program, attempt to parse it as a pattern *)
+  match parse_pattern_opt lang text with
+  | Some _ -> true
+  | None -> false
+
+(******************************************************************************)
+(* Entry Point *)
+(******************************************************************************)
+
+(* Attempts to render a fix. If unsuccessful, returns None.
+ *
+ * Failure causes include, but are not limited to:
+ * - The fix pattern does not parse.
+ * - A metavariable is bound to an AST node that is not suitable for the context
+ *   in which it is used in the fix.
+ * - Printing of the resulting fix AST fails (probably because there is simply a
+ *   node that is unhandled).
+ * *)
+let render_fix lang metavars ~fix_pattern ~target_contents =
+  (* Fixes are not exactly patterns, but they can contain metavariables that
+   * should be substituted with the nodes to which they are bound in the match.
+   * Because they can contain metavariables, we need to parse them as patterns.
+   * *)
+  let* fix_pattern_ast = parse_pattern_opt lang fix_pattern in
+
+  (* Look through the pattern and replace metavariables with the nodes to which
+   * they are bound in the match. *)
+  let* fixed_ast =
+    Autofix_metavar_replacement.replace_metavars metavars fix_pattern_ast
+  in
```

Comment:
Why having this intermediate step where you replace in the fix_pattern_ast the metavar with their value?
Why not in Autofix_printer.print_ast() pass down the metavars, and as you encounter a node
that is a metavar, just look it up and get its corresponding text from the target file (without having to pretty/ugly print anything)?

Reply:
If we did that, we'd be pretty much back to where we are with textual autofix. In the example given in https://github.com/returntocorp/semgrep/issues/4964, we would encounter the `$...AFTER` argument node while printing, and print the empty string. But, we wouldn't know without a special check to see if it was bound to zero arguments, so we would print the extraneous comma just like we currently do. We could start adding checks like that wherever we think they are needed to the printing code, but I think it's much cleaner to first create a valid generic AST and then print it. Plus, I suspect that as we improve autofix, we will want to do more operations over the resulting AST.

I will add a comment explaining this.

Reply:
I'm adding some more explanation here. Given the short length of the `render_fix` function, I'd rather explain inline than add a long file header that people are less likely to read.